test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -652,6 +652,73 @@ public void Set_connection_string_to_empty()
         Assert.That(() => conn.Open(), Throws.Exception.TypeOf<InvalidOperationException>());
     }
 
+    [Test]
+    [TestCase("test_schema_1", "public", true)]
+    [TestCase("test_schema_1", "test_schema_2", true)]
+    [TestCase("test_schema_3", "test_schema_4", true)]
+    [TestCase("test_schema_1", "public", false)]
+    [TestCase("test_schema_1", "test_schema_2", false)]
+    [TestCase("test_schema_3", "test_schema_4", false)]
+    [TestCase("'DROP TABLE X", "'COMMIT;  ", false)]
+    [Parallelizable(ParallelScope.None)]
+    public async Task Set_Schemas_And_Load_Relevant_Types(string testSchema, string otherSchema, bool enabled)
+    {
+        if (IsMultiplexing)
+            return;
+
+        using var conn1 = new NpgsqlConnection(ConnectionString);
+
+        await conn1.OpenAsync();
+        await conn1.ExecuteNonQueryAsync("DROP TYPE IF EXISTS public.test_type_3");
+        await conn1.ExecuteNonQueryAsync("CREATE TYPE public.test_type_3 AS (id int, name text)");
+
+        if (testSchema != "public")
+        {
+            await conn1.ExecuteNonQueryAsync($"DROP SCHEMA IF EXISTS \"{testSchema}\" CASCADE");
+            await conn1.ExecuteNonQueryAsync($"CREATE SCHEMA \"{testSchema}\"");
+        }
+        if (otherSchema != "public")
+        {
+            await conn1.ExecuteNonQueryAsync($"DROP SCHEMA IF EXISTS \"{otherSchema}\" CASCADE");
+            await conn1.ExecuteNonQueryAsync($"CREATE SCHEMA \"{otherSchema}\"");
+        }
+        await conn1.ExecuteNonQueryAsync($"DROP TYPE IF EXISTS \"{testSchema}\".test_type_1");
+        await conn1.ExecuteNonQueryAsync($"CREATE TYPE \"{testSchema}\".test_type_1 AS (id int)");
+        await conn1.ExecuteNonQueryAsync($"DROP TYPE IF EXISTS \"{otherSchema}\".test_type_2");
+        await conn1.ExecuteNonQueryAsync($"CREATE TYPE \"{otherSchema}\".test_type_2 AS (id int, name text)");
+        await conn1.DisposeAsync();
+
+        using var dataSource = CreateDataSource(builder =>
+        {
+            builder.ConfigureTypeLoading(builder =>
+            {
+                if (enabled)
+                    builder.SetTypeLoadingSchemas(testSchema, otherSchema);
+            });
+        });
+        using var conn = await dataSource.OpenConnectionAsync();
+        if (enabled)
+        {
+            Assert.True(dataSource.DatabaseInfo.CompositeTypes.Any(x => x.Name == "test_type_1"));
+            if (testSchema == "public" || otherSchema == "public")
+            {
+                Assert.True(dataSource.DatabaseInfo.CompositeTypes.Any(x => x.Name == "test_type_2"));
```

Comment:
We can move that above to be together with the check for `test_type_1`.

Reply:
I personally like to have the Asserts be separate so it's easier to find which assert failed. (the multiple lookups here will be against a very small set, we don't load the table composites)